Free Listings + Paid Ads: Add the billing setup card and the paid ads completion

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implements the **📌 Billing Ads card** in #1610.

- Change to open a resizable popup window for setting up the billing of Google Ads account.
- Add an alternative link for opening the billing setup page
- Add a card for notifying the successfully added billing method.
- Complete the onboarding with paid ads setup.

### Screenshots:

https://user-images.githubusercontent.com/17420811/190395908-89480b8d-896b-4ea3-ae80-4b0970ed5042.mp4

### Detailed test instructions:

1. Preparation:
   - Disconnect Google Ads account
   - Delete some GLA settings to enter the onboarding flow
      ```sql
      DELETE FROM wp_options WHERE option_name = 'gla_mc_setup_completed_at';
      ```
2. Go to the last step of onboarding flow.
3. Connect to or Create a Google Ads account 
4. Check if the UI is similar to the visual design on Figma.
5. Click on the "click here instead" link should open Google's ads billing setup page on a new browser tab.
6. Click on the "Set up billing" button should open Google's ads billing setup page with a popup window.
   - The window should be resizable.
   - The window should vertically and horizontally align to the center of the current page.
7. Finish the billing setup and go back to the onboarding flow. Check if the statue is changed to a success notice after a while.
   - To skip adding real billing info, add the code snippet to the bottom of google-listings-and-ads.php on your local site:
      ```php
      add_filter(
      	'woocommerce_gla_ads_billing_setup_status',
      	function( $status ) {
      		return 'approved';
      	}
      );
      ```
8. Enter a campaign budget and finish the onboarding with the "Complete setup" button.
9. After redirecting to the Product Feed page, navigate to the Dashboard page to see whether the just-submitted paid campaign has been created.

💡 Since the states of Google Ads connection and its billing setup are standalone and may be on a middle status, it would be nice if try to refresh at any time to see if the UI could resume from a middle status. Please note that other runtime states such as the state of "show paid ads setup sections" will be handled in a separate PR later.

### Changelog entry

eason9487 committed:
Add Google Ads account status to constants.js

eason9487 committed:
Invalidate the store state of billing status when receiving Google Ads account data

eason9487 committed:
Move files to the shared components directory

eason9487 committed:
Add the hasFinishedResolution state to useGoogleAdsAccountBillingStatus hook

eason9487 committed:
Implement the BillingCard component

eason9487 committed:
Adjust the billing setup UI to the new design and open the setup URL via a popup window

eason9487 committed:
Add the billing card to the last step of onboarding flow

eason9487 committed:
Implement the compaign creation and paid ads completion for the onboarding flow

eason9487 commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
```diff
@@ -24,37 +27,68 @@ import PaidAdsFeaturesSection from './paid-ads-features-section';
 
-	const handleChange = ( _, values, isValid ) => {
-		onCampaignChange( { ...values, isValid } );
+	const initialValues = {
+		amount: 0,
+		countryCodes: targetAudience,
 	};
+	const [ campaign, setCampaign ] = useState( {
+		...initialValues,
+		isValid: ! Object.keys( validateForm( initialValues ) ).length,
+		isReady: false,
+	} );
```
📝 
The initial values processing here aren't applied to Form `initialValues` below is temporary and will be changed in a separate PR, which implements the states storing and restoring for the Form.

https://github.com/woocommerce/google-listings-and-ads/blob/174b70a56d9e819f71c4a52c34c12baddd83a533/js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js#L77-L80

puntope commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
```diff
@@ -24,37 +27,68 @@ import PaidAdsFeaturesSection from './paid-ads-features-section';
+	// continue the setup from a middle state. For example, refresh the current page
+	// after the billing setup is finished.
+	useEffect( () => {
+		onStatesReceivedRef.current( {
```
❓  why we need to use REF here and not just `onStatesReceived` prop?

puntope replied:
❓  Also could you elaborate more the comment added here? 

I understand that we are trying to prevent some kind of issue in regards Step 4??? But not sure if I understand it fully.

eason9487 replied:
> why we need to use REF here and not just `onStatesReceived` prop?

The `onStatesReceived` callback is not a wanted deps for triggering this useEffect but it needs to be invoked with the latest reference within this useEffect.

eason9487 replied:
> Also could you elaborate more the comment added here?
> 
> I understand that we are trying to prevent some kind of issue in regards Step 4??? But not sure if I understand it fully.

I rephrased the code comment in https://github.com/woocommerce/google-listings-and-ads/pull/1677/commits/bda14476da754ecdd2af833603a9ccc7ea4766e9. Let me know if there is anything still confusing or unclear.

puntope commented on js/src/hooks/useGoogleAdsAccountBillingStatus.js:
```diff
@@ -8,14 +8,18 @@ import { useSelect } from '@wordpress/data';
+
 const useGoogleAdsAccountBillingStatus = () => {
 	return useSelect( ( select ) => {
-		const billingStatus = select(
-			STORE_KEY
-		).getGoogleAdsAccountBillingStatus();
+		const selector = select( STORE_KEY );
 
 		return {
-			billingStatus,
+			billingStatus: selector[ selectorName ](),
+			hasFinishedResolution: selector.hasFinishedResolution(
+				selectorName,
+				[]
+			),
```
❓  What is exactly this change for? I guess it's for refactor `selectorName`  and not write twice `getGoogleAdsAccountBillingStatus` but I'm not sure

eason9487 replied:
> What is exactly this change for?

To add the `hasFinishedResolution` property to the returning object. It's used [here](https://github.com/woocommerce/google-listings-and-ads/blob/ae4682c0f68206f73f36ac70dbad83e0d41cb56d/js/src/components/paid-ads/billing-card/billing-card.js#L25-L32).

> I guess it's for refactor `selectorName` and not write twice `getGoogleAdsAccountBillingStatus` but I'm not sure

Creating a variable for the selector name was based on a past code review https://github.com/woocommerce/google-listings-and-ads/pull/377#discussion_r601730266 to prevent from receiving the same suggestion. In this case, while I thought/think both ways are fine, and there is no consistency in the current codebase for similar hooks such as [useGoogleMCAccount](https://github.com/woocommerce/google-listings-and-ads/blob/ae4682c0f68206f73f36ac70dbad83e0d41cb56d/js/src/hooks/useGoogleMCAccount.js#L37-L41), I don't have an obvious preference. So, I'm happy to adjust if you have a more obvious preference for it.

puntope replied:
Nono, It's ok. Was just curious if was just for that variable name or something more. Im fine also for both.

puntope commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
```diff
@@ -24,37 +27,68 @@ import PaidAdsFeaturesSection from './paid-ads-features-section';
-const { CONNECTED } = GOOGLE_ADS_ACCOUNT_STATUS;
-
-function PaidAdsSectionsGroup( { onCampaignChange } ) {
+function PaidAdsSectionsGroup( { onStatesReceived } ) {
```
💅  WDYT of moving this to a separate component inside `setup-paid-ads-folder`

puntope replied:
💅  WDYT of naming this component `CreatePaidAdsCampaign` ? I agree that this is a group of Sections related to Paid Ads... but `PaidAdsSectionsGroup` was not very clarifying for me in terms of what action we are doing and what those "Sections groups" are doing .

eason9487 replied:
To facilitate the new round of code reviews, I will merge the component extraction to this PR from the `tweak/1610-extract-paid-ads-setup-sections` branch after this PR gets approval. The diff can be viewed in this link - https://github.com/woocommerce/google-listings-and-ads/compare/ae4682c0f68206f73f36ac70dbad83e0d41cb56d...79fcd9d9d70b8c7cb5196a4e7f9d94d92cb51f32

puntope commented on js/src/components/paid-ads/billing-card/billing-card.js:
```diff
@@ -0,0 +1,53 @@
+	}
+
+	return (
+		<BillingSetupCard
```
❓ I see `onSetupComplete` is mandatory in  `<BillingSetupCard>` but somehow it's not provided here?

eason9487 replied:
It's actually an optional prop. Updated in https://github.com/woocommerce/google-listings-and-ads/pull/1677/commits/316a7341bd21aac475e87703e47f076c6e40cd6a.

puntope commented:
💅 ❓ 

I expected some Unit tests at least for the new components. Not sure if it will be done in further PR or they are just missed

puntope commented:
> Click on the "Set up billing" button should open Google's ads billing setup page with a popup window.
The window should be resizable.
The window should vertically and horizontally align to the center of the current page.

For me it opens This documentation page (fallbackUrl) https://support.google.com/google-ads/answer/2375375

puntope commented:
In regards design. 

I miss the billing Ad in green after GoogleAdAccount Card

<img width="937" alt="Screenshot 2022-09-19 at 20 39 33" src="https://user-images.githubusercontent.com/5908855/191068825-f1ced01d-f2fc-4325-a167-b25f747e7e8f.png">

💅  Also maybe to consider tweak TreeSelectComponent padding/size. This could be done in further PR. I'm also having same issue in GLA Attributes (it also uses Tree Select at some point)

puntope commented:
❓  Not sure if this will be done in this PR or in the future.

But I still see The "Create Ads Campaign" Slide after completing the setup...

I guess in case the user set up Ads... this slide shouldn't be there anymore?

eason9487 committed:
Rephrase the code comment for more details of billing status watching in the onboarding flow.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1677#discussion_r974394966

eason9487 committed:
Change the onSetupComplete callback to an optional prop for BillingSetupCard since it's actually optional.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1677#discussion_r974435881

eason9487 committed:
Fix a typo for a code comment in SetupPaidAds component

eason9487 committed:
Extract PaidAdsSectionsGroup as a standalone component PaidAdsSetupSections from SetupPaidAds.

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/1677#discussion_r974388734
- https://github.com/woocommerce/google-listings-and-ads/pull/1677#discussion_r974392405

eason9487 commented:
Hi @puntope, thanks for the review.

> I expected some Unit tests at least for the new components. Not sure if it will be done in further PR or they are just missed

I will check if there are useful tests that could be added after the priority targets in this project are completed.

---

> For me it opens This documentation page (fallbackUrl) https://support.google.com/google-ads/answer/2375375

Please refer to this code comment:

https://github.com/woocommerce/google-listings-and-ads/blob/ae4682c0f68206f73f36ac70dbad83e0d41cb56d/js/src/components/paid-ads/billing-card/fallbackBillingUrl.js#L1-L14

---

> In regards design.
> 
> I miss the billing Ad in green after GoogleAdAccount Card
> 
> <img alt="Screenshot 2022-09-19 at 20 39 33" width="937" src="https://user-images.githubusercontent.com/5908855/191068825-f1ced01d-f2fc-4325-a167-b25f747e7e8f.png">

It appears conditionally when the connected Google Ads account is eligible for free ad credit.

https://github.com/woocommerce/google-listings-and-ads/blob/ae4682c0f68206f73f36ac70dbad83e0d41cb56d/js/src/hooks/useFreeAdCredit.js#L16-L20

---

> Not sure if this will be done in this PR or in the future.
> 
> But I still see The "Create Ads Campaign" Slide after completing the setup...
> 
> I guess in case the user set up Ads... this slide shouldn't be there anymore?

Sorry, I'm not sure about what it means. Could you clarify what it refers to or perhaps provide a screenshot?

puntope commented:
> Sorry, I'm not sure about what it means. Could you clarify what it refers to or perhaps provide a screenshot?

In the design this Card after completing the setup only show the congratulations slide

<img width="814" alt="Screenshot 2022-09-20 at 10 34 14" src="https://user-images.githubusercontent.com/5908855/191184464-69d3cdea-4140-4b64-9bbd-2abe4a0f9f4f.png">

However, in the implementation, we still show a second slide in that cards suggestion to create an Ads Campaign

I guess it should be there if we created and Ads Campaign already in setup, right?

eason9487 commented:
@puntope, thanks for the clarification.

> In the design this Card after completing the setup only show the congratulations slide
> However, in the implementation, we still show a second slide in that cards suggestion to create an Ads Campaign
> I guess it should be there if we created and Ads Campaign already in setup, right?

Yes, it needs some accordingly changes and was implemented in #1679. This PR only focuses on the billing card and the completion of paid ads setup.

puntope reviewed (approved):
✅  LGTM

eason9487 commented:
Quoted from https://github.com/woocommerce/google-listings-and-ads/pull/1677#discussion_r974878117:

> To facilitate the new round of code reviews, I will merge the component extraction to this PR from the `tweak/1610-extract-paid-ads-setup-sections` branch after this PR gets approval.

Merged in https://github.com/woocommerce/google-listings-and-ads/commit/79fcd9d9d70b8c7cb5196a4e7f9d94d92cb51f32
